snuba_sdk/query_visitors.py

```diff
@@ -354,6 +359,23 @@ def _combine(self, query: "main.Query", returns: Mapping[str, None]) -> None:
                     raise InvalidQuery(
                         f"outer query is referencing column {c.name} that does not exist in subquery"
                     )
+        elif isinstance(query.match, Join):
```

Comment:
Her I would reiterate my comment I provided in the PR below this one. 
While the difference between simple query, composite and join is just 20 lines each, the issue is the cognitive overload of having three responsibilities in this same class, which in this case are three different modes of functioning. 
Being in the same class the separation between these functions is less clear than it could be.
By splitting them instead, the structure of the code itself immediately tells who takes care of what and that there are three different ways to validate. This is useful for people who read this code for the first time and are less familiar with the code base.

Reply:
I disagree. I don't think a single if statement is harder to read than multiple separate classes. Particularly because there is only this one small difference between the classes. I think you could make a case that it's easier to read, since all the logic is in one place instead of being spread around.

Reply:
Mh, let's find a compromise.

I believe you are right tin saying that one single branch in the behavior of a class is not really harmful. It becomes more concerning when we add more. Then the responsibilities of the class increase exponentially. At that point, usually, if we are not super disciplined, we do not refactor the class because it is too complex.

Would you agree with that point? If yes and we commit in doing such refactoring when needed, then there would not be any issue with this PR, and I will accept.

Reply:
I definitely agree. And I am definitely open to doing the refactoring. As I mentioned before, I was already considering doing it for this change. Any more divergence would probably warrant a splitting of logic.